query-planner-wasm/src/lib.rs

```diff
@@ -4,31 +4,63 @@ use apollo_query_planner::{QueryPlanner, QueryPlanningOptions};
 use js_sys::JsString;
 use wasm_bindgen::prelude::*;
 
-static mut SCHEMA: Vec<String> = vec![];
-static mut DATA: Vec<QueryPlanner> = vec![];
+/// These vectors serve as repositories for the Schema string and QueryPlanner
+/// so that they don't get garbage collected. The JavaScript side gets the index
+/// into the vector, which it can use in `getQueryPlan`.
+static mut SCHEMA: Vec<Option<String>> = vec![];
+static mut DATA: Vec<Option<QueryPlanner>> = vec![];
 
+/// getQueryPlanner creates a QueryPlanner if needed, and returns an "id"
+/// for later use with `getQueryPlan`. Calling this multiple times with
+/// the same schema will only result in the schema being parsed once, and the
+/// QueryPlanner is cloned.
 #[wasm_bindgen(js_name = getQueryPlanner)]
 pub fn get_query_planner(schema: JsString) -> usize {
+    let schema = String::from(schema);
     unsafe {
-        if SCHEMA.is_empty() {
-            SCHEMA.push(String::from(schema));
-            DATA.push(QueryPlanner::new(&SCHEMA[0]));
-        } else {
-            SCHEMA[0] = String::from(schema);
-            DATA[0] = QueryPlanner::new(&SCHEMA[0]);
+        for i in 0..SCHEMA.len() {
+            match &SCHEMA[i] {
+                Some(x) if x == &schema => {
+                    let id = SCHEMA.len();
+                    SCHEMA.push(Some(schema));
+                    // No need to re-parse, we can just clone!
+                    DATA.push(DATA[i].clone());
```

Comment:
I see why you went with this approach, 2 calls using the same index and one of them dropping the planner would cause a dangling reference to an index.

### Why not clone?

We've been intentional about not implementing `Clone` for structs that can potentially be very large. For the `QueryPlanner` case, its data is the parsed schema ast, which can be a very large deeply nested tree. One of the design tenets for this gateway is to be able to operate over very large schemas (e.g. imagine an entire enterprise's data graph, composed from 100+ different services) and cloning that would create a noticeable performance degradation.

Cloning here creates a state where the gateway will be susceptible to undue memory pressure, if a user mis-uses the js library or even in the normal case, if there is a schema composition event but no change in the resulting composed schema, and we keep cloning the same planner over and over.

Generally speaking, using an `Rc` would solve that, but there's a bug here....

### Bug: `QueryPlanner` references the schema `String`.

This is the definition:
```rust
pub struct QueryPlanner<'s> {
    pub schema: schema::Document<'s>,
}
```

that `'s` is the lifetime referenced by `Document`, which, if you look at where that ends up coming from, is the lifetime of the original `&str` that contains the schema. So when we put the schema in `SCHEMA`, we then use a reference to it to pass to the parser. By cloning (even if this was an `Rc`), the cloned planner still references the `String` that was used to create the original planner. When dropping that planner and its schema, any cloned planner will have dangling references.

### How should we fix the root issue?

We've not had a chance to discuss this, and we appreciate the contribution and ideas, I think they may inform an eventual solution. Ideally I'd like to figure out if we can fix the problem outside of the surface area of these objects somehow.

### Idea?

Just had a thought, maybe we should be adding a static `GATEWAYS` vector that will contain a schema and planner, which would map to a js `ApolloGateway` instance?
Meaning, if you have 2 `ApolloGateway` instances, each of which is constantly replacing its schema, you'd have 2 items in a `GATEWAYS` vector, and for each gateway we'd still just have one schema and planner that's being replaced?
In that situation we'd not have to implement any `drop` (which really adds a lot of complexity)

That might be a better design. Doing that would mean introducing some breaking API changes, the names would match what we're returning:

```rust
#[wasm_bindgen(js_name = getGatewayBridge)]
pub fn get_gateway_bridge(schema: JsString) -> usize { .. }

#[wasm_bindgen(js_name = getQueryPlan)]
pub fn get_query_plan(gateway_idx: usize, query: &str, options: &JsValue) -> JsValue { .. }
```

(Noting that this would be a breaking API change, which will be relevant when releasing)